### ISSUE TITLE
append says that [] isn't an array, but %() is

### ISSUE THREAD

**Issue by pdebelak** [oilshell/oil#1252]
I was trying to learn how to add an item to a list, and so I ran `oil -c 'help append'` after looking at the tour of oil. The output is:

```
$ oil -c 'help append'
 append 

Append a string to an array of strings:
var mylist = %(one two)
push :mylist three

This is a command-mode synonym for the expression:
_ mylist.append('three')
```

The `_ mylist.append('three')` part is helpful, because that works, but the `push :mylist tree` part is confusing because it appears that `push` isn't part of `oil`. It also doesn't appeared to be listed as unimplemented when looking at `oil -c 'help oil'`, but maybe I'm missing it.

**Comment by andychu** [comment]
Thanks for the report, I just fixed it!      `append` used to be called `push`, but the help wasn't properly updated.

So now it's `append :mylist three`, which is a lot more consistent :)

Let me know if you have other questions!

**Comment by pdebelak** [comment]
@andychu thanks! Now I'm confused about how `append` works.

This works:

```
$ oil -c 'var output = []; for item in ~/* { _ output.append(item) }; json write (output)'
[
  "/Users/pdebelak/Desktop",
  # etc
]
```

But this doesn't:

```
$ oil -c 'var output = []; for item in ~/* { append :output item }; json write (output)'   
  var output = []; for item in ~/* { append :output item }; json write (output)
                                            ^
[ -c flag ]:1: 'output' isn't an array
  var output = []; for item in ~/* { append :output item }; json write (output)
                                     ^~~~~~
[ -c flag ]:1: errexit PID 13420: Command failed with status 1
```

I may or may not be using `append` wrong somehow in the second example, but the error message about `'output' isn't an array` doesn't seem correct and doesn't really guide me to how to fix it.

```
$ oil --version
Oil version 0.12.0
Release Date: 2022-07-24 03:56:32+00:00
Arch: arm64
OS: Darwin
Platform: Darwin Kernel Version 21.6.0: Sat Jun 18 17:07:22 PDT 2022; root:xnu-8020.140.41~1/RELEASE_ARM64_T6000
Compiler: GCC Apple LLVM 13.1.6 (clang-1316.0.21.2.5)
Interpreter: OVM
Interpreter version: 2.7.13
Bytecode: bytecode-opy.zip
```

I get the same results in linux x86_64 for oil 0.12.0 so I don't think this is a macos problem.

**Comment by andychu** [comment]
Oof, sorry that's a bug.  The fundamental issue is #636, i.e. we need to stop using the Python interpreter.   There are a few other bugs along those lines:

https://github.com/oilshell/oil/labels/bug

-----

There are two workarounds:

- Use the append() method as you noticed
- Use the `%()` literal which are unquoted words, like shell arrays, or Perl's `q`

```
$ oil -c 'var output = %(); for item in ~/* { append :output $item }; json w
rite (output)'
[
  "/home/andy/Desktop",
  "/home/andy/Documents",
  "/home/andy/Downloads",
```

The intention is that `%(foo bar)` is syntactic sugar for `['foo', 'bar']`, but that isn't quite true now unfortunately.  Thanks for the report

**Comment by andychu** [comment]
FWIW I believe this was a recent breakage due to some refactoring toward #636 -- it used to work :-/

Not sure when this can be fixed but I'll take a look

**Comment by andychu** [comment]
I made a spot fix for `append`  ... there is a more fundamental issue which needs to be addressed with  #636 though.  Thanks for the feedback!

Released: https://www.oilshell.org/release/0.12.3/
